### PR TITLE
Feature/181 Linq queries

### PR DESCRIPTION
**Pull request by vsatkov**
issue #181

### CODE REVIEW

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+
+            foreach (var item in customers)
+            {
+                ObjectDumper.Write($"City: {item.City}, Customer: {item.Name}");
+                foreach (var supl in item.Suppls)
+                {
+                    ObjectDumper.Write($"Supplier: supl");
+                }
+                ObjectDumper.Write(String.Empty);
+            }
```
Comment (vladimirchr) [inline HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: I think we need to use 
`foreach (var item in customers)`
`{`
`                ObjectDumper.Write(item , 1);`
`}`
for displaying objects. And we need to implement second part of the task: 
>Сделайте задания с использованием группировки и без.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        public void Linq6()
+        {
+            /* A4
+             * Выдайте список клиентов с указанием, начиная с какого месяца какого года они стали клиентами (принять за таковые месяц и год самого первого заказа)
+             */            
+            // 2nd ver.
+            var clients =
+                from c in dataSource.Customers
+                where c.Orders.Length > 0
+                let FirstOrder = c.Orders.Take(1)
```
Comment (vladimirchr) [inline HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: We dont use this variable.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+                select new
+                {
+                    Name = c.CompanyName,
+                    StartYear = c.Orders.Take(1).FirstOrDefault().OrderDate.Year,
+                    StartMonth = c.Orders.Take(1).FirstOrDefault().OrderDate.Month
```
Comment (vladimirchr) [inline HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: We can skip `.Take(1)` part if we using `.FirstOrDefault()`

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+            /* A9
+             * Рассчитайте среднюю прибыльность каждого города (среднюю сумму заказа по всем клиентам из данного города) и среднюю интенсивность (среднее количество заказов, приходящееся на клиента из каждого города)
+             */
+            var cities = dataSource.Customers.
+                Select(c => new { c.City, c.Country }).Distinct().
+                Select(c => new {
+                    numberOfClients = dataSource.Customers.Where(cstmr => cstmr.City == c.City && cstmr.Country == c.Country).
+                    Select(cstmr => new {
+                        numberOfOrders = cstmr.Orders.Length,
+                        sumOrderPrices = cstmr.Orders.Select(p => p.Total).Sum(),
```
Comment (vladimirchr) [inline HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: We can use `Sum()` without `Select()`: 
`cstmr.Orders.Sum(p => p.Total)`

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+             */
+            Dictionary<int, int> monthsStatistic;
+            Dictionary<int, int> yearsStatistic;
+            Dictionary<string, int> yearsAndMonthsStatistic;
+
+            var clients = dataSource.Customers.Select(c =>
+            new {
+                c.CompanyName,
+                ordersDates = c.Orders.Select(o => o.OrderDate)
+            });
```
Comment (vladimirchr) [inline HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: I think the main idea was to teach us how to write complex Linq queries.
But your solution is working, so i think task is done)

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+
```
Comment (alch-0) [inline HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: All the logic has to be inside the query with the output in one foreach cycle.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+             * Сгруппируйте все продукты по категориям, внутри – по наличию на складе, внутри последней группы отсортируйте по стоимости
+             */
+            // 2nd ver.
+            var products =
+                from p in dataSource.Products
+                group p by p.Category into c
+                group c by c.First().UnitsInStock into s
+                group s by s.First().First().UnitPrice
+                ;
+
```
[the remark this answers is not in the document]
Comment (vsatkov) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: Thank you for your patience. I misunderstood the condition 'по наличию на складе' of this task

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+            foreach (var item in customers)
+            {
+                ObjectDumper.Write($"City: {item.City}, Customer: {item.Name}");
+                foreach (var supl in item.Suppls)
+                {
+                    ObjectDumper.Write($"Supplier: supl");
+                }
+                ObjectDumper.Write(String.Empty);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (vsatkov) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: Thank you! I didn't use all functional ObjectDumper cause I didn't know about it. It's helpful tool.

[the remark this answers is not in the document]
Comment (alch-0) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: I checked the new version.
In the task it says "For each client" so the initial grouping has to be made by client ID/client name plus client's country and city.
Inside the group there has to be names(or name) of the supplier from the same country and city

So it should look like this:
Client1 Name/Client ID            Country           City
            Supplier1 name
            Supplier2 name
Client2 Name/Client ID            Country           City
            Supplier3 name

It seems that for the data we make queries to there are no clients that have more than one supplier in their country and city, so for this data there are only one customer to one supplier pairs. But there are suppliers that are situated in the same country and city as several customers.

Also the query has to return only the data about the client-supplier match, so you need to remove the information about the clients that don't have suppliers in their cities from the query result.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (alch-0) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: Checked the new version, look great.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs (hunk 6 above)
Comment (alch-0) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: The new version is fine.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+            foreach (var group in products)
+            {
+                ObjectDumper.Write(group.Key);
+                foreach (var product in group)
+                {
+                    ObjectDumper.Write($"   {product.ProductName} — {product.UnitPrice}");
+                }
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (alch-0) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: Check the revised version, looks OK.

File: HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (vsatkov) [reply HW5/Valeriy_Satkov/Task/Task/LinqSamples.cs]: > Also the query has to return only the data about the client-supplier match, so you need to remove the information about the clients that don't have suppliers in their cities from the query result.
We have '**For each** customer, make a list of suppliers' in task that means I shouldn't delete customers without suppliers from list.

**Pull request merged**